File: typescript/libs/workspace/adapters/system/src/lib/session/ChangeSet.ts
import { IChangeSet, IStrategy, ISession } from '@allors/workspace/domain/system';
import { AssociationType, PropertyType, RelationType, RoleType } from '@allors/workspace/meta/system';
import { MapMap } from '../collections/MapMap';
import { Session } from './Session';
import { Strategy } from './Strategy';
import { frozenEmptySet } from '../collections/frozenEmptySet';
import { IRange, Ranges } from '../collections/ranges/Ranges';

export class ChangeSet implements IChangeSet {
  associationsByRoleType: Map<RoleType, Set<IStrategy>>;
  rolesByAssociationType: Map<AssociationType, Set<IStrategy>>;

  private ranges: Ranges<Strategy>;

  public constructor(public session: Session, public created: Set<IStrategy>, public instantiated: Set<IStrategy>) {
    this.associationsByRoleType = new Map();
    this.rolesByAssociationType = new Map();

    this.created ??= frozenEmptySet as Set<IStrategy>;
    this.instantiated ??= frozenEmptySet as Set<IStrategy>;

    this.ranges = this.session.ranges;
  }

  public addSessionStateChanges(sessionStateChangeSet: MapMap<PropertyType, Strategy, unknown>) {
    for (const [propertyType, map] of sessionStateChangeSet.mapMap) {
      const strategies = new Set<IStrategy>();

      for (const [strategy] of map) {
        strategies.add(strategy);
      }

      if (propertyType.isAssociationType) {
        this.rolesByAssociationType.set(propertyType as AssociationType, strategies);
      } else if (propertyType.isRoleType) {
        this.associationsByRoleType.set(propertyType as RoleType, strategies);
      } else {
        throw new Error(`PropertyType ${propertyType.name} is not supported`);
      }
    }
  }

  public diffUnit(association: Strategy, relationType: RelationType, current: unknown, previous: unknown) {
    if (current !== previous) {
      this.addAssociation(relationType, association);
    }
  }

  public diffCompositeStrategyRecord(association: Strategy, relationType: RelationType, current: Strategy, previous: number) {
    if (current?.id === previous) {
      return;
    }

    if (previous != null) {
      this.addRole(relationType, (this.session as Session).getStrategy(previous));
    }

    if (current != null) {
      this.addRole(relationType, current);
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositeRecordRecord(association: Strategy, relationType: RelationType, current: number, previous: number) {
    if (current === previous) {
      return;
    }

    if (previous != null) {
      this.addRole(relationType, (this.session as Session).getStrategy(previous));
    }

    if (current != null) {
      this.addRole(relationType, (this.session as Session).getStrategy(current));
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositeStrategyStrategy(association: Strategy, relationType: RelationType, current: Strategy, previous: Strategy) {
    if (current === previous) {
      return;
    }

    if (previous != null) {
      this.addRole(relationType, previous);
    }

    if (current != null) {
      this.addRole(relationType, current);
    }

    this.addAssociation(relationType, association);
  }

  public diffCompositesStrategyRecord(association: Strategy, relationType: RelationType, current: IRange<Strategy>, previousRange: IRange<number>) {
    const previous: IRange<Strategy> = previousRange?.map((v) => (this.session as Session).getStrategy(v));
    this.diffCompositesStrategyStrategy(association, relationType, current, previous);
  }

  public diffCompositesRecordRecord(association: Strategy, relationType: RelationType, currentRange: IRange<number>, previousRange: IRange<number>) {
    const current: IRange<Strategy> = currentRange?.map((v) => (this.session as Session).getStrategy(v));
    const previous: IRange<Strategy> = previousRange?.map((v) => (this.session as Session).getStrategy(v));
    this.diffCompositesStrategyStrategy(association, relationType, current, previous);
  }

  public diffCompositesStrategyStrategy(association: Strategy, relationType: RelationType, current: IRange<Strategy>, previous: IRange<Strategy>) {
    let hasChange = false;

    for (const v of this.ranges.enumerate(previous)) {
      if (!this.ranges.has(current, v)) {
        this.addRole(relationType, v);
        hasChange = true;
      }
    }

    for (const v of this.ranges.enumerate(current)) {
      if (!this.ranges.has(previous, v)) {
        this.addRole(relationType, v);
        hasChange = true;
      }
    }

    if (hasChange) {
      this.addAssociation(relationType, association);
    }
  }

  private addAssociation(relationType: RelationType, association: Strategy) {
    const roleType = relationType.roleType;

    let associations = this.associationsByRoleType.get(roleType);
    if (!associations) {
      associations = new Set();
      this.associationsByRoleType.set(roleType, associations);
    }

    associations.add(association);
  }

  private addRole(relationType: RelationType, role: Strategy) {
    const associationType = relationType.associationType;

    let roles = this.rolesByAssociationType.get(associationType);
    if (!roles) {
      roles = new Set();
      this.rolesByAssociationType.set(associationType, roles);
    }

    roles.add(role);
  }
}

File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters/session/ChangeSet.cs
// <copyright file="ChangeSet.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Collections;
    using Meta;
    using Ranges;

    public sealed class ChangeSet : IChangeSet
    {
        public ChangeSet(Session session, ISet<IStrategy> created, ISet<IStrategy> instantiated)
        {
            this.Session = session;
            this.Created = created ?? EmptySet<IStrategy>.Instance;
            this.Instantiated = instantiated ?? EmptySet<IStrategy>.Instance;
            this.AssociationsByRoleType = new Dictionary<IRoleType, ISet<IStrategy>>();
            this.RolesByAssociationType = new Dictionary<IAssociationType, ISet<IStrategy>>();
        }

        private Session Session { get; }

        ISession IChangeSet.Session => this.Session;

        public ISet<IStrategy> Created { get; }

        public ISet<IStrategy> Instantiated { get; }

        public IDictionary<IRoleType, ISet<IStrategy>> AssociationsByRoleType { get; }

        public IDictionary<IAssociationType, ISet<IStrategy>> RolesByAssociationType { get; }

        public void AddSessionStateChanges(IDictionary<IPropertyType, IDictionary<long, object>> sessionStateChangeSet)
        {
            foreach (var kvp in sessionStateChangeSet)
            {
                var ids = kvp.Value.Keys;
                var strategies = new HashSet<IStrategy>(ids.Select(v => this.Session.GetStrategy(v)));

                switch (kvp.Key)
                {
                    case IAssociationType associationType:
                        this.RolesByAssociationType.Add(associationType, strategies);
                        break;
                    case IRoleType roleType:
                        this.AssociationsByRoleType.Add(roleType, strategies);
                        break;
                    default:
                        throw new ArgumentOutOfRangeException($"PropertyType {kvp.Key}");
                }
            }
        }

        public void DiffUnit(Strategy association, IRelationType relationType, object current, object previous)
        {
            if (!Equals(current, previous))
            {
                this.AddAssociation(relationType, association);
            }
        }

        public void DiffComposite(Strategy association, IRelationType relationType, Strategy current, long? previous)
        {
            if (Equals(current?.Id, previous))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, this.Session.GetStrategy((long)previous));
            }

            if (current != null)
            {
                this.AddRole(relationType, current);
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposite(Strategy association, IRelationType relationType, long? current, Strategy previous)
        {
            if (Equals(current, previous?.Id))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, previous);
            }

            if (current != null)
            {
                this.AddRole(relationType, this.Session.GetStrategy((long)current));
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposite(Strategy association, IRelationType relationType, Strategy current, Strategy previous)
        {
            if (Equals(current, previous))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, previous);
            }

            if (current != null)
            {
                this.AddRole(relationType, current);
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposites(Strategy association, IRelationType relationType, ISet<Strategy> current, IRange previousRange)
        {
            var ranges = this.Session.Workspace.Ranges;
            var previous = new HashSet<Strategy>(ranges.Ensure(previousRange).Select(v => this.Session.GetStrategy(v)));
            this.DiffComposites(association, relationType, current, previous);
        }

        public void DiffComposites(Strategy association, IRelationType relationType, IRange currentRange, ISet<Strategy> previous)
        {
            var ranges = this.Session.Workspace.Ranges;
            var current = new HashSet<Strategy>(ranges.Ensure(currentRange).Select(v => this.Session.GetStrategy(v)));
            this.DiffComposites(association, relationType, current, previous);
        }

        public void DiffComposites(Strategy association, IRelationType relationType, ISet<Strategy> current, ISet<Strategy> previous)
        {
            var hasChange = false;

            foreach (var v in current.Except(previous))
            {
                this.AddRole(relationType, v);
                hasChange = true;
            }

            foreach (var v in previous.Except(current))
            {
                this.AddRole(relationType, v);
                hasChange = true;
            }

            if (hasChange)
            {
                this.AddAssociation(relationType, association);
            }
        }

        private void AddAssociation(IRelationType relationType, Strategy association)
        {
            var roleType = relationType.RoleType;

            if (!this.AssociationsByRoleType.TryGetValue(roleType, out var associations))
            {
                associations = new HashSet<IStrategy>();
                this.AssociationsByRoleType.Add(roleType, associations);
            }

            associations.Add(association);
        }

        private void AddRole(IRelationType relationType, Strategy role)
        {
            var associationType = relationType.AssociationType;

            if (!this.RolesByAssociationType.TryGetValue(associationType, out var roles))
            {
                roles = new HashSet<IStrategy>();
                this.RolesByAssociationType.Add(associationType, roles);
            }

            roles.Add(role);
        }
    }
}
